src/Features/CSharp/Portable/CodeRefactorings/InlineMethod/CSharpInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,244 @@
+                if (arrowExpressionNode != null)
+                {
+                    inlineSyntaxNode = arrowExpressionNode.Expression;
+                }
+
+                return inlineSyntaxNode;
+            }
+
+            // A check has been done before to make sure there is one arrow expression or block statement.
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        private static SyntaxNode? GetExpressionFromStatementSyntaxNode(StatementSyntax statementSyntax)
+            => statementSyntax switch
+            {
+                ReturnStatementSyntax returnStatementSyntax => returnStatementSyntax.Expression,
+                ExpressionStatementSyntax expressionStatementSyntax => expressionStatementSyntax.Expression,
+                _ => null
+            };
+
+        protected override bool ShouldCheckTheExpressionPrecedenceInCallee(SyntaxNode syntaxNode)
+            => s_syntaxKindsNeedsToCheckThePrecedence.Any(syntaxNode.IsKind);
+
+        protected override bool NeedWrapInParenthesisWhenPrecedenceAreEqual(SyntaxNode calleeInvocationSyntaxNode)
+        {
+            var parent = calleeInvocationSyntaxNode.Parent;
+            if (parent != null)
+            {
+                // For left associative expression
+                // If in the original invocation, it is the left child. Since it is a left associative expression,
+                // then it is safe to replace it without parenthesis.
+                // Example:
+                // int Callee(int i, int j) => i + j;
+                // void Caller()
+                // {
+                //   var a = F(1, 2) - 3 - 2 - 1;
+                // }
+                if (s_leftAssociativeSyntaxKinds.Any(parent.IsKind) && parent is BinaryExpressionSyntax leftAssociativeBinaryExpressionSyntax)
+                {
+                    return leftAssociativeBinaryExpressionSyntax.Left != calleeInvocationSyntaxNode;
+                }
+
+                // Same for right associative expression, if the original invocation is the right child,
+                // it is safe to replace it.
+                if (parent.IsKind(SyntaxKind.CoalesceExpression)
+                    && parent is BinaryExpressionSyntax rightAssociativeBinaryExpressionSyntax)
+                {
+                    return rightAssociativeBinaryExpressionSyntax.Right != calleeInvocationSyntaxNode;
+                }
+                else if (parent is ConditionalExpressionSyntax conditionalExpressionSyntax)
+                {
+                    return conditionalExpressionSyntax.WhenFalse != calleeInvocationSyntaxNode;
+                }
+                else if (parent is AssignmentExpressionSyntax assignmentExpressionSyntax)
+                {
+                    return assignmentExpressionSyntax.Right != calleeInvocationSyntaxNode;
+                }
+            }
+
+            // In other cases, always put a 'safe' parenthesis around
+            return true;
+        }
```

Comment:
this feels super complex.  Why not just always paranthesize.  The simplifier will then remove the parens if unnecessary.

Reply:
Do you mean always add the parenthesis and later let the user to remove the parethesis by using the code action? Or if there is an existing parenthesis service that can do the clean up before the document is returned?

Reply:
the latter.  just use the `.Parenthesize()` extension method and it should do the work for you.

Reply:
Done, that's a great helper utilities method